File: typescript/tests/hosting_service.test.ts
// hosting_service.test.ts
//
// Verifies CircleAI.Hosting.AIService: startup via an injected generator
// factory, ask/chat, system-prompt enrichment (device context + persona +
// skills + RAG), streaming with observer events, tool-call parsing + the
// agentic loop, feedback-driven persona adaptation, and the RT-04 brownout
// hot-swap through a fallback-chain selector.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  AIService,
  AIObserverBase,
  type AIChatEvent,
  type IFallbackChainModelSelector,
  type AIOptions,
} from "../src/hosting/index";
import type {
  GenerationOptions,
  IChatGenerator,
  ModelSelection,
} from "../src/inference/index";
import type { ChatMessage } from "../src/models/index";
import type { DeviceProbe } from "../src/device/index";
import { NullDeviceContext } from "../src/device/index";
import {
  InMemoryPersonaStore,
  InMemoryFeedbackStore,
  InMemoryEpisodicStore,
} from "../src/memory/stores";
import { FeedbackPolarity, type FeedbackSignal } from "../src/memory/index";
import { InMemorySkillStore } from "../src/hosting/index";

/** Records the messages each call received and replies with a scripted string. */
class ScriptedGenerator implements IChatGenerator {
  lastMessages: readonly ChatMessage[] = [];
  calls = 0;
  disposed = false;
  constructor(private readonly replies: string[] | ((n: number) => string)) {}

  private reply(): string {
    if (typeof this.replies === "function") return this.replies(this.calls);
    return this.replies[Math.min(this.calls, this.replies.length - 1)];
  }

  async generateAsync(messages: readonly ChatMessage[]): Promise<string> {
    this.lastMessages = messages;
    const r = this.reply();
    this.calls++;
    return r;
  }

  async *streamAsync(messages: readonly ChatMessage[]): AsyncGenerator<string> {
    this.lastMessages = messages;
    const r = this.reply();
    this.calls++;
    for (const word of r.split(" ")) yield word;
  }

  dispose(): void {
    this.disposed = true;
  }
}

function optionsWith(over: Partial<AIOptions>): AIOptions {
  return { modelPath: "/fake/model.gguf", warmOnStart: false, ...over };
}

describe("AIService — startup + ask/chat", () => {
  it("loads via the injected factory and answers a question", async () => {
    const gen = new ScriptedGenerator(["hi there"]);
    const svc = new AIService(optionsWith({ generatorFactory: () => gen }));
    assert.equal(svc.isReady, false);
    await svc.startAsync();
    assert.equal(svc.isReady, true);
    const answer = await svc.askAsync("hello");
    assert.equal(answer, "hi there");
    await svc.disposeAsync();
    assert.equal(gen.disposed, true);
  });

  it("throws when neither a model path nor a factory can resolve a model", async () => {
    const svc = new AIService({ warmOnStart: false });
    await assert.rejects(() => svc.startAsync());
  });

  it("prepends an enriched system prompt with device context", async () => {
    const gen = new ScriptedGenerator(["ok"]);
    const ctx = {
      ...new NullDeviceContext(),
      activeAppId: "com.example.app",
      networkType: "wifi",
      batteryLevel: 0.5,
      isCharging: true,
    };
    const svc = new AIService(
      optionsWith({ generatorFactory: () => gen, deviceContext: ctx }),
    );
    await svc.chatAsync([{ role: "user", content: "hey" }]);
    const sys = gen.lastMessages.find((m) => m.role === "system");
    assert.ok(sys, "a system message should be injected");
    assert.match(sys!.content, /\[Device context\]/);
    assert.match(sys!.content, /Active app: com.example.app/);
    assert.match(sys!.content, /Battery: 50% \(charging\)/);
  });

  it("honours a caller-supplied system message verbatim (no enrichment)", async () => {
    const gen = new ScriptedGenerator(["ok"]);
    const ctx = { ...new NullDeviceContext(), activeAppId: "app" };
    const svc = new AIService(
      optionsWith({ generatorFactory: () => gen, deviceContext: ctx }),
    );
    await svc.chatAsync([
      { role: "system", content: "CUSTOM" },
      { role: "user", content: "hi" },
    ]);
    const systems = gen.lastMessages.filter((m) => m.role === "system");
    assert.equal(systems.length, 1);
    assert.equal(systems[0].content, "CUSTOM");
  });

  it("injects skill context when a skill store is configured", async () => {
    const gen = new ScriptedGenerator(["ok"]);
    const skills = new InMemorySkillStore();
    await skills.upsertAsync("summarise", {
      name: "Summarise",
      description: "Summarise long text",
      instructions: "Be concise.",
      tags: ["text"],
    });
    const svc = new AIService(
      optionsWith({ generatorFactory: () => gen, skillStore: skills }),
    );
    await svc.chatAsync([{ role: "user", content: "please summarise this" }]);
    const sys = gen.lastMessages.find((m) => m.role === "system");
    assert.match(sys!.content, /## Available Skills/);
    assert.match(sys!.content, /\*\*summarise\*\*/);
  });
});

describe("AIService — streaming + observer", () => {
  it("streams pieces and fires stream started/completed", async () => {
    const gen = new ScriptedGenerator(["one two three"]);
    const events: string[] = [];
    let completedTokens = -1;
    const observer = new (class extends AIObserverBase {
      override async onStreamStartedAsync() {
        events.push("start");
      }
      override async onStreamCompletedAsync(e: { tokenCount: number }) {
        events.push("complete");
        completedTokens = e.tokenCount;
      }
    })();
    const svc = new AIService(optionsWith({ generatorFactory: () => gen, observer }));
    const out: string[] = [];
    for await (const p of svc.streamAsync([{ role: "user", content: "go" }])) out.push(p);
    assert.deepEqual(out, ["one", "two", "three"]);
    assert.deepEqual(events, ["start", "complete"]);
    assert.equal(completedTokens, 3);
  });

  it("stores the exchange in episodic memory and fires onChatCompleted", async () => {
    const gen = new ScriptedGenerator(["stored"]);
    const episodic = new InMemoryEpisodicStore();
    const captured: AIChatEvent[] = [];
    const observer = new (class extends AIObserverBase {
      override async onChatCompletedAsync(e: AIChatEvent) {
        captured.push(e);
      }
    })();
    const svc = new AIService(
      optionsWith({ generatorFactory: () => gen, episodicMemory: episodic, observer }),
    );
    await svc.chatAsync([{ role: "user", content: "remember me" }]);
    assert.equal(await episodic.countAsync(), 1);
    assert.equal(captured.length, 1);
    assert.equal(captured[0].response, "stored");
  });
});

describe("AIService — tool parsing + agentic loop", () => {
  it("parseToolCall extracts name + arguments from the Qwen tag", () => {
    const inv = AIService.parseToolCall(
      'sure <tool_call>{"name":"weather","arguments":{"city":"Durban","days":3}}</tool_call>',
    );
    assert.ok(inv);
    assert.equal(inv!.toolName, "weather");
    assert.equal(inv!.arguments["city"], "Durban");
    // Non-string args become raw JSON text (matches the C# ToManaged path).
    assert.equal(inv!.arguments["days"], "3");
  });

  it("returns null when no tool call is present", () => {
    assert.equal(AIService.parseToolCall("just text"), null);
  });

  it("agentic loop invokes the tool bridge then re-prompts to a plain answer", async () => {
    // First turn emits a tool call; second turn returns plain text.
    const gen = new ScriptedGenerator((n) =>
      n === 0
        ? '<tool_call>{"name":"echo","arguments":{"v":"hi"}}</tool_call>'
        : "final answer",
    );
    const bridge = {
      availableTools: [],
      async invoke(inv: { toolName: string; arguments: Record<string, unknown> }) {
        return { toolName: inv.toolName, success: true, result: inv.arguments["v"] };
      },
    };
    const svc = new AIService(
      optionsWith({ generatorFactory: () => gen, toolBridge: bridge }),
    );
    const result = await svc.agenticChatAsync("do it");
    assert.equal(result, "final answer");
    assert.equal(gen.calls, 2);
  });

  it("invokeTool returns a failure when no bridge is configured", async () => {
    const gen = new ScriptedGenerator(["x"]);
    const svc = new AIService(optionsWith({ generatorFactory: () => gen }));
    const res = await svc.invokeToolAsync({ toolName: "t", arguments: {} });
    assert.equal(res.success, false);
    assert.match(res.error ?? "", /No tool bridge configured/);
  });
});

describe("AIService — feedback persona adaptation", () => {
  it("negative feedback nudges verbosity toward brief and persists", async () => {
    const gen = new ScriptedGenerator(["x"]);
    const personaStore = new InMemoryPersonaStore();
    const feedbackStore = new InMemoryFeedbackStore();
    const svc = new AIService(
      optionsWith({ generatorFactory: () => gen, personaStore, feedbackStore }),
    );

    // 10 negative signals → analyser drives verbosity down.
    for (let i = 0; i < 10; i++) {
      const sig: FeedbackSignal = {
        id: `s${i}`,
        recordedAtUtc: new Date(),
        userText: "u",
        assistantText: "a",
        polarity: FeedbackPolarity.Negative,
      };
      await svc.submitFeedbackAsync(sig);
    }

    const persona = await personaStore.loadAsync("default");
    assert.equal(persona.verbosity, "brief");
    assert.equal(persona.negativeSignals, 10);
    assert.equal(await feedbackStore.countAsync(), 10);
  });
});

describe("AIService — RT-04 brownout", () => {
  it("hot-swaps to the next model in the fallback chain", async () => {
    const made: string[] = [];
    const loader = {
      async downloadModelAsync(id: string) {
        return `/models/${id}.gguf`;
      },
      getModelPath(id: string) {
        return `/models/${id}.gguf`;
      },
      async modelExists() {
        return true;
      },
      async checkForCriticalUpdateAsync() {
        return false;
      },
      dispose() {},
    };
    const selector: IFallbackChainModelSelector = {
      bestFit(_probe: DeviceProbe): ModelSelection {
        return {
          modelId: "big",
          requiresDownload: false,
          estimatedBytes: 0,
          tier: 3 as ModelSelection["tier"],
        };
      },
      allCandidates() {
        return [];
      },
      chainFor(id: string) {
        return id === "big" ? ["big", "small"] : ["small"];
      },
    };
    const gen = () => {
      const g = new ScriptedGenerator(["x"]);
      return g;
    };
    let swappedTo = "";
    const observer = new (class extends AIObserverBase {
      override async onBrownoutAsync(_from: string, to: string) {
        swappedTo = to;
      }
    })();
    const svc = new AIService(
      { modelId: "big", warmOnStart: false, observer },
      loader,
      gen,
      selector,
    );
    await svc.startAsync();
    void made;
    const ok = await svc.brownoutAsync(3 /* Manual */);
    assert.equal(ok, true);
    assert.equal(swappedTo, "small");
    await svc.disposeAsync();
  });

  it("brownout is a no-op without a chainFor-capable selector", async () => {
    const gen = new ScriptedGenerator(["x"]);
    const svc = new AIService(optionsWith({ generatorFactory: () => gen }));
    await svc.startAsync();
    const ok = await svc.brownoutAsync(0);
    assert.equal(ok, false);
    await svc.disposeAsync();
  });
});

// GenerationOptions is referenced only for typing; silence unused-import lints.
export type _Options = GenerationOptions;

File: rust/tests/hosting_service_test.rs
//! hosting_service_test.rs
//!
//! Verifies the AIService lifecycle + chat/stream/agentic/feedback flows and the
//! FallbackAIService RAM-gated backend selection. Mirrors the C# AIService /
//! FallbackAIService behaviours.

use std::sync::Mutex;

use circle_ai::hosting::service::{
    AIOptions, AIService, DeviceContext, FallbackAIService, FixedRamProbe, HostingError,
    IAIService, IHostChatGenerator, IHostToolBridge,
};
use circle_ai::inference::{ChatMessage, GenerationOptions};
use circle_ai::tools::{ToolInvocation, ToolResult};

/// Deterministic host generator: echoes the last user message with a prefix,
/// or emits a queued canned reply (for agentic tool-call scripting).
struct ScriptGenerator {
    replies: Mutex<std::collections::VecDeque<String>>,
    default_reply: String,
}

impl ScriptGenerator {
    fn new(default_reply: &str) -> Self {
        Self {
            replies: Mutex::new(std::collections::VecDeque::new()),
            default_reply: default_reply.to_string(),
        }
    }
    fn push(&self, reply: &str) {
        self.replies.lock().unwrap().push_back(reply.to_string());
    }
}

impl IHostChatGenerator for ScriptGenerator {
    fn generate(
        &self,
        _messages: &[ChatMessage],
        _options: Option<&GenerationOptions>,
    ) -> Result<String, String> {
        Ok(self
            .replies
            .lock()
            .unwrap()
            .pop_front()
            .unwrap_or_else(|| self.default_reply.clone()))
    }
    fn stream(
        &self,
        _messages: &[ChatMessage],
        _options: Option<&GenerationOptions>,
    ) -> Result<Vec<String>, String> {
        // Two chunks so token_count = 2.
        Ok(vec!["Hello ".to_string(), "world".to_string()])
    }
}

struct EchoToolBridge;
impl IHostToolBridge for EchoToolBridge {
    fn invoke(&self, invocation: &ToolInvocation) -> ToolResult {
        ToolResult::ok(&invocation.tool_name, Some(serde_json::json!({ "ok": true })))
    }
}

fn base_options() -> AIOptions {
    AIOptions {
        warm_on_start: false,
        ..AIOptions::default()
    }
}

#[test]
fn start_is_idempotent_and_sets_ready() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("hi")));
    assert!(!svc.is_ready());
    svc.start().unwrap();
    assert!(svc.is_ready());
    // Second start is a no-op.
    svc.start().unwrap();
    assert!(svc.is_ready());
}

#[test]
fn ask_prepends_enriched_system_prompt_and_returns_reply() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("the answer")));
    let out = svc.ask("what is 2+2?").unwrap();
    assert_eq!(out, "the answer");
}

#[test]
fn chat_auto_starts_when_not_started() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("reply")));
    // chat without an explicit start() — ensure_started kicks in.
    let out = svc
        .chat(&[ChatMessage::user("hi")], None)
        .unwrap();
    assert_eq!(out, "reply");
    assert!(svc.is_ready());
}

#[test]
fn stream_returns_all_chunks() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("x")));
    svc.start().unwrap();
    let chunks = svc.stream(&[ChatMessage::user("hi")], None).unwrap();
    assert_eq!(chunks, vec!["Hello ".to_string(), "world".to_string()]);
}

#[test]
fn invoke_tool_without_bridge_returns_failure() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("x")));
    let mut args = std::collections::HashMap::new();
    args.insert("a".to_string(), serde_json::json!(1));
    let result = svc.invoke_tool(&ToolInvocation::new("t", args)).unwrap();
    assert!(!result.success);
    assert_eq!(result.error.as_deref(), Some("No tool bridge configured."));
}

#[test]
fn invoke_tool_with_bridge_succeeds() {
    let opts = AIOptions {
        warm_on_start: false,
        tool_bridge: Some(Box::new(EchoToolBridge)),
        ..AIOptions::default()
    };
    let svc = AIService::new(opts, Box::new(ScriptGenerator::new("x")));
    let result = svc
        .invoke_tool(&ToolInvocation::new("echo", std::collections::HashMap::new()))
        .unwrap();
    assert!(result.success);
}

#[test]
fn agentic_chat_runs_tool_then_returns_plaintext() {
    let gen = ScriptGenerator::new("done");
    // First reply asks for a tool; second is plain text and terminates the loop.
    gen.push(r#"<tool_call>{"name":"echo","arguments":{"x":1}}</tool_call>"#);
    gen.push("all done");
    let opts = AIOptions {
        warm_on_start: false,
        tool_bridge: Some(Box::new(EchoToolBridge)),
        agentic_max_iterations: Some(5),
        ..AIOptions::default()
    };
    let svc = AIService::new(opts, Box::new(gen));
    let out = svc.agentic_chat("please echo", None).unwrap();
    assert_eq!(out, "all done");
}

#[test]
fn parse_tool_call_supports_tool_name_key() {
    let inv = AIService::parse_tool_call(
        r#"noise <tool_call>{"tool_name":"gmail.send","arguments":{"to":"a@b.c"}}</tool_call> trailing"#,
    )
    .unwrap();
    assert_eq!(inv.tool_name, "gmail.send");
    assert_eq!(inv.arguments.get("to").unwrap(), &serde_json::json!("a@b.c"));
}

#[test]
fn parse_tool_call_returns_none_without_tags() {
    assert!(AIService::parse_tool_call("just some text").is_none());
}

#[test]
fn disposed_semantics_via_stop_then_start_again() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("x")));
    svc.start().unwrap();
    svc.stop().unwrap();
    assert!(!svc.is_ready());
    // Re-start works (stop does not dispose).
    svc.start().unwrap();
    assert!(svc.is_ready());
}

#[test]
fn ask_empty_question_errors() {
    let svc = AIService::new(base_options(), Box::new(ScriptGenerator::new("x")));
    let err = svc.ask("").unwrap_err();
    assert_eq!(err, HostingError::Failed("question required".to_string()));
}

#[test]
fn device_context_enriches_prompt_but_reply_unchanged() {
    // Reply is fixed; this asserts device-context enrichment does not error and
    // the flow still returns the generator's reply.
    let opts = AIOptions {
        warm_on_start: false,
        device_context: Some(DeviceContext {
            local_time: Some("2026-07-08 12:00".to_string()),
            battery_level: Some(0.55),
            is_charging: Some(true),
            network_type: Some("wifi".to_string()),
            active_app_id: Some("bruh".to_string()),
            ..DeviceContext::default()
        }),
        ..AIOptions::default()
    };
    let svc = AIService::new(opts, Box::new(ScriptGenerator::new("ok")));
    assert_eq!(svc.ask("hi").unwrap(), "ok");
}

// ── FallbackAIService ───────────────────────────────────────────────────────

fn make_service(reply: &str) -> Box<dyn IAIService> {
    Box::new(AIService::new(
        base_options(),
        Box::new(ScriptGenerator::new(reply)),
    ))
}

#[test]
fn fallback_uses_local_when_ram_above_threshold() {
    let fb = FallbackAIService::new(
        make_service("local"),
        make_service("cloud"),
        Some(1000),
        Box::new(FixedRamProbe(2000)),
    );
    fb.start().unwrap();
    assert!(fb.is_local_active());
    assert_eq!(fb.ask("q").unwrap(), "local");
}

#[test]
fn fallback_uses_cloud_when_ram_below_threshold() {
    let fb = FallbackAIService::new(
        make_service("local"),
        make_service("cloud"),
        Some(4000),
        Box::new(FixedRamProbe(1000)),
    );
    fb.start().unwrap();
    assert!(fb.is_cloud_active());
    assert_eq!(fb.ask("q").unwrap(), "cloud");
}

#[test]
fn fallback_errors_before_start() {
    let fb = FallbackAIService::new(
        make_service("local"),
        make_service("cloud"),
        None,
        Box::new(FixedRamProbe(0)),
    );
    assert!(fb.ask("q").is_err());
}
